controllers/release/release_adapter_test.go

```diff
@@ -0,0 +1,421 @@
+		adapter = NewAdapter(release, ctrl.Log, k8sClient, ctx)
+		Expect(reflect.TypeOf(adapter)).To(Equal(reflect.TypeOf(&Adapter{})))
+	})
+
+	AfterEach(func() {
+		err := k8sClient.Delete(ctx, release)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+	})
+
+	AfterAll(func() {
+		err := k8sClient.Delete(ctx, releaseLink)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+
+		err = k8sClient.Delete(ctx, applicationSnapshot)
+		Expect(err == nil || errors.IsNotFound(err)).To(BeTrue())
+
+		Expect(k8sClient.Delete(ctx, releaseStrategy)).Should(Succeed())
+	})
+
+	It("can create a new Adapter instance", func() {
+		Expect(reflect.TypeOf(NewAdapter(release, ctrl.Log, k8sClient, ctx))).
+			To(Equal(reflect.TypeOf(&Adapter{})))
+	})
+
+	It("ensures finalizer is added to the Release object", func() {
+		result, err := adapter.EnsureFinalizerIsAdded()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		Expect(release.ObjectMeta.GetFinalizers()[0]).
+			To(ContainSubstring("release-finalizer"))
+
+		result, err = adapter.EnsureFinalizerIsAdded()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+	})
+
+	It("ensures a PipelineRun object exists", func() {
+		// some racing condition in EnvTest makes it throw an error even if the CancelRequest is false
+		// so I am adding a function to wait for it to succeed before continuing.
+		Eventually(func() bool {
+			result, err := adapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}, time.Second*10).Should(BeTrue())
+
+		pipelineRun, _ := adapter.getReleasePipelineRun()
+		Expect(reflect.TypeOf(pipelineRun)).
+			To(Equal(reflect.TypeOf(&tektonv1beta1.PipelineRun{})))
+		Expect(pipelineRun).NotTo(BeNil())
+
+		// clean up
+		Expect(k8sClient.Delete(ctx, pipelineRun)).Should(Succeed())
+	})
+
+	It("ensures the Release Pipeline is tracked", func() {
+		Eventually(func() bool {
+			result, err := adapter.EnsureReleasePipelineRunExists()
+			return (!result.CancelRequest && err == nil)
+		}, time.Second*10).Should(BeTrue())
+		pipelineRun, _ := adapter.getReleasePipelineRun()
+		Expect(pipelineRun).NotTo(BeNil())
+
+		result, err := adapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		// should return results.ContinueProcessing() in case the release is done
+		release.MarkSucceeded()
+		result, err = adapter.EnsureReleasePipelineStatusIsTracked()
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
+
+		// release is running but has no pipelineRun matching the release
+		release.MarkRunning()
+		pipelineRun.ObjectMeta.Labels = nil
+		k8sClient.Update(ctx, pipelineRun)
+		release.Status.Conditions = []metav1.Condition{}
+		result, err = adapter.EnsureReleasePipelineStatusIsTracked()
+
+		// err should not be nil here but adapter.getReleasePipelineRun does not advise
+		// there is no pipelineRun for the release
+		Expect(!result.CancelRequest && err == nil).To(BeTrue())
```

Comment:
The check should probably be:
```
Expect(!result.RequeueRequest && pipelineRun == nil).To(BeTrue())
```
I'd also remove the comment.

Reply:
I changed this to another Eventually block getting the `pipelineRun` and error is nil.